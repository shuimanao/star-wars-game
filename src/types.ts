export type AppProps = {};

export type GameState =
  | { status: 'initial' }
  | { status: 'in_progress'; chosenResource: 'people' | 'starships' };

export type SwapiResult = Record<string, string>[];

export type SwapiResponse = {
  count: number;
  next: string;
  previous: string;
  results: SwapiResult;
};

export type PlayerProps = {
  data: SwapiResult;
  setGameState: (arg: GameState) => void;
};

export type Person = {
  name: string;
  height: string;
  mass: string;
  hair_color: string;
  skin_color: string;
  eye_color: string;
  birth_year: string;
  gender: string;
  homeworld: string;
  films: string;
  species: string;
  vehicles: string;
  starships: string;
  created: string;
  edited: string;
  url: string;
};

export type Starship = {
  MGLT: string;
  cargo_capacity: string;
  consumables: string;
  cost_in_credits: string;
  created: string;
  crew: string;
  edited: string;
  films: string;
  hyperdrive_rating: string;
  length: string;
  manufacturer: string;
  max_atmosphering_speed: string;
  model: string;
  name: string;
  passengers: string;
  pilots: string;
  starship_class: string;
  url: string;
};
