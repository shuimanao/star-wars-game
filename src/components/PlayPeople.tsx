import React, { FC, useState } from 'react';
import { Button } from '@mui/material';
import { choosePlayers } from '../utils/choosePlayers';
import { renderChosenPlayers } from '../utils/renderChosenPlayers';
import { renderWinner } from '../utils/renderWinner';
import { PlayerProps, SwapiResult } from '../types';

export const PlayPeople: FC<PlayerProps> = ({ data, setGameState }) => {
  const [chosenPlayers, setChosenPlayers] = useState<SwapiResult>([]);
  const [numberOfGames, setNumberOfGames] = useState(0);
  const resource = 'people';

  const handlePlayButton = () => {
    const chosen = choosePlayers(data);
    setChosenPlayers(chosen);
    setNumberOfGames(numberOfGames + 1);
  };

  const handleResetGameButton = () => setGameState({ status: 'initial' });

  return (
    <>
      <h4 data-testid="h2eople-chosen">You've chosen People to play</h4>
      {renderChosenPlayers(chosenPlayers, resource)}
      {renderWinner(chosenPlayers, resource)}
      {numberOfGames === 0 && <p>Click the button to choose two cards:</p>}
      <Button
        onClick={handlePlayButton}
        data-testid="play-btn-people"
        variant="contained"
      >
        {numberOfGames > 0 ? <>Play again</> : <>Play</>}
      </Button>
      <Button onClick={handleResetGameButton} variant="outlined">
        Reset game
      </Button>
    </>
  );
};
