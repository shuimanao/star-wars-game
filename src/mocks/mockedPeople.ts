import { Person } from '../types';

export const mockedPeople: Person[] = [
  {
    name: 'Leia Organa',
    height: '150',
    mass: '49',
    hair_color: 'brown',
    skin_color: 'light',
    eye_color: 'brown',
    birth_year: '19BBY',
    gender: 'female',
    homeworld: 'https://swapi.py4e.com/api/planets/2/',
    films: '',
    species: '',
    vehicles: '',
    starships: '',
    created: '2014-12-10T15:20:09.791000Z',
    edited: '2014-12-20T21:17:50.315000Z',
    url: 'https://swapi.py4e.com/api/people/5/',
  },
  {
    name: 'R5-D4',
    height: '97',
    mass: '32',
    hair_color: 'n/a',
    skin_color: 'white, red',
    eye_color: 'red',
    birth_year: 'unknown',
    gender: 'n/a',
    homeworld: 'https://swapi.py4e.com/api/planets/1/',
    films: '',
    species: '',
    vehicles: '',
    starships: '',
    created: '2014-12-10T15:57:50.959000Z',
    edited: '2014-12-20T21:17:50.321000Z',
    url: 'https://swapi.py4e.com/api/people/8/',
  },
];
