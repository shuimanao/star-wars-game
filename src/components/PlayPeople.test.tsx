import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/dom';
import '@testing-library/jest-dom';
import { PlayPeople } from './PlayPeople';
import { PlayerProps } from '../types';
import { mockedPeople } from '../mocks/mockedPeople';

const mockedProps: PlayerProps = {
  data: mockedPeople,
  setGameState: jest.fn(),
};

test('<PlayPeople /> renders correctly', () => {
  const { container } = render(<PlayPeople {...mockedProps} />);
  expect(container).toMatchSnapshot();

  const element = screen.getByTestId('play-btn-people');
  userEvent.click(element);

  expect(screen.queryAllByTestId('people-card')).toBeTruthy();
});
