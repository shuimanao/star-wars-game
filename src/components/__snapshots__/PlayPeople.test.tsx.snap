// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayPeople /> renders correctly 1`] = `
<div>
  <h4
    data-testid="h2eople-chosen"
  >
    You've chosen People to play
  </h4>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  />
  <p>
    There is no winner
  </p>
  <p>
    Click the button to choose two cards:
  </p>
  <button
    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
    data-testid="play-btn-people"
    tabindex="0"
    type="button"
  >
    Play
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Reset game
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
